source/common/buffer/buffer_impl.h

```diff
@@ -1,16 +1,427 @@
+    uint64_t copy_size;
+    if (dataSize() == 0) {
+      // There is nothing in the slice, so put the data at the very end in case the caller
+      // later tries to prepend anything else in front of it.
+      copy_size = std::min(size, reservableSize());
+      reservable_ = capacity_;
+      data_ = capacity_ - copy_size;
+    } else {
+      if (data_ == 0) {
+        // There is content in the slice, and no space in front of it to write anything.
+        return 0;
+      }
+      // Write into the space in front of the slice's current content.
+      copy_size = std::min(size, data_);
+      data_ -= copy_size;
+    }
+    memcpy(base_ + data_, src + size - copy_size, copy_size);
+    return copy_size;
+  }
+
+protected:
+  Slice(uint64_t data, uint64_t reservable, uint64_t capacity)
+      : data_(data), reservable_(reservable), capacity_(capacity) {}
+
+  /** Start of the slice - subclasses must set this */
+  uint8_t* base_{nullptr};
+
+  /** Offset in bytes from the start of the slice to the start of the Data section */
+  uint64_t data_;
+
+  /** Offset in bytes from the start of the slice to the start of the Reservable section */
+  uint64_t reservable_;
+
+  /** Total number of bytes in the slice */
+  uint64_t capacity_;
+
+  /** Whether reserve() has been called without a corresponding commit(). */
+  bool reservation_outstanding_{false};
+};
+
+using SlicePtr = std::unique_ptr<Slice>;
+
+class OwnedSlice : public Slice {
+public:
+  /**
+   * Create an empty OwnedSlice.
+   * @param capacity number of bytes of space the slice should have.
+   * @return an OwnedSlice with at least the specified capacity.
+   */
+  static SlicePtr create(uint64_t capacity) {
+    uint64_t slice_capacity = sliceSize(capacity);
+    return SlicePtr(new (slice_capacity) OwnedSlice(slice_capacity));
+  }
+
+  /**
+   * Create an OwnedSlice and initialize it with a copy of the supplied copy.
+   * @param data the content to copy into the slice.
+   * @param size length of the content.
+   * @return an OwnedSlice containing a copy of the content, which may (dependent on
+   *         the internal implementation) have a nonzero amount of reservable space at the end.
+   */
+  static SlicePtr create(const void* data, uint64_t size) {
+    uint64_t slice_capacity = sliceSize(size);
+    std::unique_ptr<OwnedSlice> slice(new (slice_capacity) OwnedSlice(slice_capacity));
+    memcpy(slice->base_, data, size);
+    slice->reservable_ = size;
+    return slice;
+  }
+
+  // Custom delete operator to keep C++14 from using the global operator delete(void*, size_t),
+  // which would result in the compiler error:
+  // "exception cleanup for this placement new selects non-placement operator delete"
+  static void operator delete(void* address) { ::operator delete(address); }
+
+private:
+  static void* operator new(size_t object_size, size_t data_size) {
+    return ::operator new(object_size + data_size);
```

Comment:
nit: align object_size before adding it to data_size?

Reply:
`::operator new` guarantees that its return value will be "suitably aligned to hold an object of any fundamental alignment": https://en.cppreference.com/w/cpp/memory/new/operator_new

Reply:
I know that ... I was thinking that adding object_size and data_size without aligning either might  result in an allocation size that assumed a packing that would not correspond to the structure offsets. But now I can't think of a scenario where two realistic sizes would combine into something that wouldn't be aligned when passed to new(). So I guess this is OK.

Sorry I'm just being paranoid :)